enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/log/segmented/ReaderPool.java

```diff
@@ -0,0 +1,129 @@
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.coreedge.raft.log.segmented;
+
+import java.io.IOException;
+import java.time.Clock;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import org.neo4j.io.fs.FileSystemAbstraction;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.LogProvider;
+
+// TODO: Prune stale readers with a recurring job.
+class ReaderPool
+{
+    private final ArrayList<Reader> pool;
+    private final int maxSize;
+    private Log log;
+    private final FileNames fileNames;
+    private final FileSystemAbstraction fsa;
+    private final Clock clock;
+
+    ReaderPool( int maxSize, LogProvider logProvider, FileNames fileNames, FileSystemAbstraction fsa, Clock clock )
+    {
+        this.pool = new ArrayList<>( maxSize );
+        this.maxSize = maxSize;
+        this.log = logProvider.getLog( getClass() );
+        this.fileNames = fileNames;
+        this.fsa = fsa;
+        this.clock = clock;
+    }
+
+    Reader acquire( long version, long byteOffset ) throws IOException
+    {
+        Reader reader = getFromPool( version );
+        if ( reader == null )
+        {
+            reader = createFor( version );
+        }
+
+        reader.channel().position( byteOffset );
+        return reader;
+    }
+
+    void release( Reader reader )
+    {
+        reader.setTimeStamp( clock.millis() );
+        Optional<Reader> optionalOverflow = putInPool( reader );
+        optionalOverflow.ifPresent( this::dispose );
+    }
+
+    private synchronized Reader getFromPool( long version )
+    {
+        Iterator<Reader> itr = pool.iterator();
+        while ( itr.hasNext() )
+        {
+            Reader reader = itr.next();
+            if ( reader.version() == version )
+            {
+                itr.remove();
+                return reader;
+            }
+        }
+        return null;
+    }
+
+    private synchronized Optional<Reader> putInPool( Reader reader )
+    {
+        Reader overflow = null;
+
+        pool.add( reader );
+        if ( pool.size() > maxSize )
+        {
+            overflow = pool.remove( 0 );
+        }
+
+        return Optional.ofNullable( overflow );
```

Comment:
Fine but it is inconsinstent with `acquire`/`getFromPool` which force you to do a null-check. Can we at least make it consistent and use Optional everywhere?